c2rust-analyze/src/type_desc.rs

```diff
@@ -94,3 +105,169 @@ pub fn local_perms_to_desc<'tcx>(
+                steps.push(Step::Ref(mutbl));
+                cur_ty = inner_ty;
+            }
+            TyKind::RawPtr(tm) => {
+                steps.push(Step::RawPtr(tm.mutbl));
+                cur_ty = tm.ty;
+            }
+            TyKind::Adt(adt_def, substs) if adt_def.is_box() => {
+                steps.push(Step::Box);
+                cur_ty = substs.type_at(0);
+            }
+            TyKind::Adt(adt_def, substs) if is_rc(tcx, adt_def) => {
+                steps.push(Step::Rc);
+                cur_ty = substs.type_at(0);
+            }
+            TyKind::Adt(adt_def, substs) if is_cell(tcx, adt_def) => {
+                steps.push(Step::Cell);
+                cur_ty = substs.type_at(0);
+            }
+            TyKind::Adt(adt_def, substs) if is_offset_ptr(tcx, adt_def) => {
+                steps.push(Step::OffsetPtr);
+                cur_ty = substs.type_at(0);
+            }
+
+            TyKind::Slice(inner_ty) => {
+                steps.push(Step::Slice);
+                cur_ty = inner_ty;
+            }
+            TyKind::Array(inner_ty, _) => {
+                steps.push(Step::Array);
+                cur_ty = inner_ty;
+            }
+
+            _ => panic!(
+                "failed to deconstruct {:?} as a pointer to {:?}: unexpected {:?}",
+                ty,
+                pointee_ty,
+                cur_ty.kind()
+            ),
+        }
+    }
+
+    let mut i = 0;
+    let next = |i: &mut usize| -> Option<&Step> {
+        let s = steps.get(*i);
+        *i += 1;
+        s
+    };
+
+    let own = match next(&mut i) {
+        Some(&Step::Ref(Mutability::Not)) => {
+            if let Some(&Step::Cell) = steps.get(i) {
+                i += 1;
+                Ownership::Cell
+            } else {
+                Ownership::Imm
+            }
+        }
+        Some(&Step::Ref(Mutability::Mut)) => Ownership::Mut,
+        Some(&Step::RawPtr(Mutability::Not)) => Ownership::Raw,
+        Some(&Step::RawPtr(Mutability::Mut)) => Ownership::RawMut,
+        Some(&Step::Box) => Ownership::Box,
+        Some(&Step::Rc) => Ownership::Rc,
+        _ => {
+            // Un-consume this step.
+            i -= 1;
+            panic!(
+                "failed to deconstruct {:?} as a pointer to {:?}: \
+                    steps {:?} don't start with a pointer",
+                ty,
+                pointee_ty,
+                &steps[i - 1..]
+            );
+        }
+    };
+
+    let qty = match next(&mut i) {
+        Some(&Step::Slice) => Quantity::Slice,
+        Some(&Step::OffsetPtr) => Quantity::OffsetPtr,
+        Some(&Step::Array) => Quantity::Array,
+        _ => {
+            // Un-consume this step.
+            i -= 1;
+            Quantity::Single
+        }
+    };
+
+    assert!(
+        i == steps.len(),
+        "failed to deconstruct {:?} as a pointer to {:?}: got extra steps {:?}",
+        ty,
+        pointee_ty,
+        &steps[i..],
+    );
```

Comment:
```suggestion
    struct Cursor<'a, T> {
        a: &'a [T],
        i: usize,
    }

    impl<'a, T> Cursor<'a, T> {
        pub fn new(a: &'a [T]) -> Self {
            Self { a, i: 0 }
        }

        pub fn next(&mut self) -> Option<&'a T> {
            let e = self.a.get(self.i);
            self.i += 1;
            e
        }

        pub fn prev(&mut self) {
            self.i -= 1;
        }

        pub fn remaining(&self) -> &'a [T] {
            &self.a[self.i..]
        }
    }

    let mut steps = Cursor::new(&steps);

    let own = match steps.next().copied() {
        Some(Step::Ref(Mutability::Not)) => {
            if let Some(Step::Cell) = steps.next().copied() {
                Ownership::Cell
            } else {
                steps.prev();
                Ownership::Imm
            }
        }
        Some(Step::Ref(Mutability::Mut)) => Ownership::Mut,
        Some(Step::RawPtr(Mutability::Not)) => Ownership::Raw,
        Some(Step::RawPtr(Mutability::Mut)) => Ownership::RawMut,
        Some(Step::Box) => Ownership::Box,
        Some(Step::Rc) => Ownership::Rc,
        _ => {
            steps.prev();
            panic!(
                "failed to deconstruct {:?} as a pointer to {:?}: \
                    steps {:?} don't start with a pointer",
                ty,
                pointee_ty,
                steps.remaining(),
            );
        }
    };

    let qty = match steps.next().copied() {
        Some(Step::Slice) => Quantity::Slice,
        Some(Step::OffsetPtr) => Quantity::OffsetPtr,
        Some(Step::Array) => Quantity::Array,
        _ => {
            steps.prev();
            Quantity::Single
        }
    };

    assert!(
        steps.remaining().is_empty(),
        "failed to deconstruct {:?} as a pointer to {:?}: got extra steps {:?}",
        ty,
        pointee_ty,
        steps.remaining(),
    );
```

Reply:
i think these diffs are pretty huge and hard to know what changed without scrutiny that surpasses whats needed for a normal pr diff, id recommend trying to make it clearer somehow what the change is

Reply:
The diff isn't quite as large as GitHub shows, as it shows unchanged lines in the diff, too.  The main change was to add `Cursor` as an easier to understand abstraction than a free `i` variable with some closures, which was much less clear to me.  This way the usages of `steps.{next, prev, remaining}()` are a lot more self-explanatory and it's clearer how the steps work.

Reply:
This part is essentially a parser, so I made it look more like one, with a single `eat` operation: eb55a007ad1080578f4958ef3d2c425421e8e7bc.  The backward stepping (`i -= 1`) was a hack to avoid repeating `i += 1` in almost every branch.